Add unified config API & SDK API

masnax opened the pull request:
Closes #280 
Closes #274 
Closes #277 

* Adds `/1.0/system/security` With `GET`, `PUT`, and `PATCH` supported.
  * `GET` will return the current system config (with private key omitted)
  * `PUT` will replace the entire system config
  * `PATCH` will update the system config with new values only

* Certificates are now included in the system config, and are configurable over the API, as are all OIDC/OpenFGA fields, as well as the listener & worker addresses. 

* It is not possible to update the config while in an active migration as it may interfere with running workers.

* Adds `/1.0/system/sdks/{sourceType}` with `GET` and `POST` supported.
  * `GET` will return the current sdk file for a given source, or `NOT FOUND`. It will return a 500 if it detects an incomplete SDK import.
  * `POST` writes the SDK file for a given source type (only vmware right now). The SDK filename is determined by the server so each POST will overwrite the file for now. 

* Adds a `config` CLI command set to manage all of the above

* Moves the expected location of the worker binary & UI directory over to `/usr/lib/migration-manager` or `/var/lib/migration-manager/static` if `MIGRATION_MANAGER_DIR` is set.

masnax commented:
Sigh, messed up the rebase from #289. Only change is the `Makefile` fix at the end.

breml commented on internal/server/sys/os.go:
```diff
@@ -8,17 +8,26 @@ import (
 	LogDir   string // Log directory (e.g. /var/log/).
 	RunDir   string // Runtime directory (e.g. /run/migration-manager/).
 	VarDir   string // Data directory (e.g. /var/lib/migration-manager/).
+	UsrDir   string // Static directory (e.g. /usr/lib/migration-manager/).
```
Based on comment https://github.com/FuturFusion/operations-center/pull/205#issuecomment-3225125447 from @stgraber, this should be changed to `/usr/share/migration-manager/ui`.

breml commented on cmd/migration-managerd/internal/api/api_system.go:
```diff
@@ -0,0 +1,235 @@
+		if replace {
+			d.config.SystemConfigPut = cfg
+		} else {
+			if cfg.RestWorkerEndpoint != "" {
```
The way `PATCH` is implemented, it only allows to set a value, that has not been set previously or change an existing value (update). It is not possible to "unset" a certain value, correct? Is this on purpose?

masnax replied:
Yeah, I can't think of a simple way to sent a set of keys to be unset, since this uses the zero-values to represent ignored keys. We would need to use null values instead, but encoding null values in a JSON would get convoluted. Maybe you could use query args, but I think that couples the Go code with the JSON tags a bit too much.

breml replied:
From a REST API point of view, I think that for the `PATCH` method it is natural to only provide the keys in the request, where a change is required and that providing a key, but setting it to `null`, does imply, that this key should be unset.

I wonder, if it is worth the price to maintain a separate struct, only used for unmarshal of the request of a  `PATCH` request, where all the fields are pointers. This way, `PATCH` could distinguish between a field not being present in the request (and those not changed), set to zero value (unsetting the field) or set to a value.

I think, we could have a unit test, which ensures, that the two structs stay in sync e.g. with some reflect code, which iterates both structs and fails, if there is not a pointer field for each field present in the reference struct from `shared/api`.

masnax replied:
In the end, it required 4 structs (OIDC config, OpenFGA config, parent Security config, and network config) but with the included reflect test, it should be fine.

masnax commented on cmd/migration-managerd/internal/api/daemon.go:
```diff
@@ -70,10 +73,12 @@ type Daemon struct {
-	config     *config.DaemonConfig
+	listener       net.Listener
+	server         *http.Server
+	serverCertLock sync.Mutex
```
Yeah, that would address the edge case of multiple requests coming in immediately after a config update request. I'm not sure about TryLock though, I'd rather leave it up to the request timeout to decide if the connection has waited long enough rather than forcing a retry, which would always take the same amount of time or more, never less.

You did draw my attention to the fact that we use the daemon's reference to the OIDC verifier multiple times when handling an API request, which may mean the verifiercan change mid-request as well.

masnax commented:
@stgraber @breml I've updated this now with the requested changes.

The config API has been split into 3 parts for network/security/certificate. Since I already had PATCH here I kept the single config.yaml file. The former two have GET/PUT/PATCH and the latter is only PUT, since a partial certificate update doesn't make much sense. 

There's also a global lock on config changes, which is shared with getter helpers for each of the handlers and the certificate, and I've made sure that within each scope, we only grab one instance of a handler. 

Also added `/usr/share/migration-manager` for the UI dir.

breml commented:
> The former two have GET/PUT/PATCH and the latter is only PUT, since a partial certificate update doesn't make much sense.

@stgraber and I discussed yesterday, that for the certificate case, it would be cleaner to use `POST` instead of `PUT`. The reasoning is, that for `PUT` we would expect to have a `GET`, which returns the same thing. This is not the case for the certificates because we do not offer `GET` and if we would offer `GET`, we would want to omit the `key`.

breml commented on shared/api/config.go:
```diff
@@ -0,0 +1,42 @@
+	OpenFGA ConfigOpenFGA `json:"openfga" yaml:"openfga"`
+}
+
+type ConfigOIDC struct {
+	Issuer   string `json:"issuer,omitempty"    yaml:"issuer,omitempty"`
+	ClientID string `json:"client.id,omitempty" yaml:"client.id,omitempty"`
+	Scope    string `json:"scopes,omitempty"    yaml:"scopes,omitempty"`
+	Audience string `json:"audience,omitempty"  yaml:"audience,omitempty"`
+	Claim    string `json:"claim,omitempty"     yaml:"claim,omitempty"`
+}
+
+type ConfigOpenFGA struct {
+	APIToken string `json:"api.token,omitempty" yaml:"api.token,omitempty"`
+	APIURL   string `json:"api.url,omitempty"   yaml:"api.url,omitempty"`
+	StoreID  string `json:"store.id,omitempty"  yaml:"store.id,omitempty"`
+}
```
I previously had the settings with a dot (`.`) as well, but now I felt, it might be clearer to have it with underscore and since the format of the config file changed in breaking ways anyway, I decided to change this as well in https://github.com/FuturFusion/operations-center/pull/208. I don't have a strong opinion on this, but I see some value in having it consistent between Migration Manager and Operations Center. What do you guys think?

stgraber replied:
Having dots in json keys definitely feels a bit odd to me, we should do underscores :)

breml commented on shared/api/config.go:
```diff
@@ -0,0 +1,42 @@
+	StoreID  string `json:"store.id,omitempty"  yaml:"store.id,omitempty"`
+}
+
+type Certificate struct {
+	Cert string `json:"cert"          yaml:"cert"`
+	CA   string `json:"ca"            yaml:"ca"`
+	Key  string `json:"key,omitempty" yaml:"key,omitempty"`
+}
```
In my case, the server certificate and key is currently in separate files (`server.crt` and `server.key`) and not part of the config file. Do you have a preference? Should we do it the same in Migration Manager and Operations Center?

breml replied:
I guess, I missed, that `Certificate` is not a field on `SystemConfig`. All good.

In my case, I named it `CertificatePost` to make it clear, that this is not data, which can be fetched.

breml commented:
> > Currently, I'm thinking:
> > 
> > * `/system/certificates` can handle the TLS certs (may add ACME there down the line)
> > * `/system/security` can handle OIDC, OpenFGA, trusted TLS certificates
> > * `/system/network` can handle the listen address/port and external URLs
> 
> With this API structure, do you also think, that the config files should be split into `config.yaml`, `security.yaml` and so on? Currently I have everything in a single config file. But since the `PUT` request is supposed to replace the entire config, I would prefer to split it into multiple files, if we go with this API structure.

Just for the record, in https://github.com/FuturFusion/operations-center/pull/208 for Operations Center I currently have a single config file. This is in line with the implementation in this PR for Migration Manager.

breml commented on cmd/migration-managerd/main_daemon.go:
```diff
@@ -49,19 +49,17 @@ func (c *cmdDaemon) Run(cmd *cobra.Command, args []string) error {
-	cfg.RestServerIPAddr = c.flagServerAddr
-	cfg.RestServerPort = c.flagServerPort
-	cfg.RestWorkerEndpoint = c.flagWorkerEndpoint
+	cfg.Network.Address = c.flagServerAddr
+	cfg.Network.Port = c.flagServerPort
+	cfg.Network.WorkerEndpoint = c.flagWorkerEndpoint
```
In regards to the validation of the configuration, this boils down to allowing `address` (Operations Center endpoint), `rest_server_address` and `rest_server_port` to be zero value and with this, the API is not served over TCP but only over the unix socket. As soon, as the `rest_server_port` or the `address` is set, the other needs to be non-zero value as well.

masnax replied:
Alright, I've made starting the http listener contingent on a config.yml being present with the relevant fields set.

masnax commented:
After discussions with @stgraber on changing the SDK API, I've removed those parts from this PR.

stgraber commented on shared/api/system.go:
```diff
@@ -7,7 +7,7 @@ type SystemConfig struct {
 type ConfigNetwork struct {
 	Address        string `json:"address,omitempty"         yaml:"address,omitempty"`
-	Port           int    `json:"port,omitempty"            yaml:"port,omitempty"`
+	Port           string `json:"port,omitempty"            yaml:"port,omitempty"`
```
I'd probably expect the API type to be an integer. The zero value would then use the default port.

masnax replied:
I thought we discussed above to not have any default port, and not start a listener at all until an address/port is set explicitly?

masnax replied:
Or is it that we just don't want to use wildcard addresses, so require an address to be set explicitly but the port can have a default?

stgraber replied:
Yeah, we shouldn't start a listener until an address has been configured.

So basically:
 - Address="" => Don't start listener (regardless of Port value)
 - Address="::" => Listens on everything, uses default port if Port is set to 0
 - Address="0.0.0.0" => Listen on all IPv4 addresses
